File: hbase-common/src/main/java/org/apache/hadoop/hbase/codec/KeyValueCodec.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.hadoop.hbase.codec;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;

import org.apache.hadoop.hbase.Cell;
import org.apache.hadoop.hbase.HBaseInterfaceAudience;
import org.apache.hadoop.hbase.KeyValue;
import org.apache.hadoop.hbase.KeyValueUtil;
import org.apache.hadoop.hbase.NoTagsKeyValue;
import org.apache.hadoop.hbase.OffheapKeyValue;
import org.apache.hadoop.hbase.ShareableMemory;
import org.apache.hadoop.hbase.classification.InterfaceAudience;
import org.apache.hadoop.hbase.nio.ByteBuff;
import org.apache.hadoop.hbase.util.ByteBufferUtils;
import org.apache.hadoop.hbase.util.Bytes;

/**
 * Codec that does KeyValue version 1 serialization.
 * 
 * <p>Encodes Cell as serialized in KeyValue with total length prefix.
 * This is how KVs were serialized in Puts, Deletes and Results pre-0.96.  Its what would
 * happen if you called the Writable#write KeyValue implementation.  This encoder will fail
 * if the passed Cell is not an old-school pre-0.96 KeyValue.  Does not copy bytes writing.
 * It just writes them direct to the passed stream.
 *
 * <p>If you wrote two KeyValues to this encoder, it would look like this in the stream:
 * <pre>
 * length-of-KeyValue1 // A java int with the length of KeyValue1 backing array
 * KeyValue1 backing array filled with a KeyValue serialized in its particular format
 * length-of-KeyValue2
 * KeyValue2 backing array
 * </pre>
 */
@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.CONFIG)
public class KeyValueCodec implements Codec {
  public static class KeyValueEncoder extends BaseEncoder {
    public KeyValueEncoder(final OutputStream out) {
      super(out);
    }

    @Override
    public void write(Cell cell) throws IOException {
      checkFlushed();
      // Do not write tags over RPC
      ByteBufferUtils.putInt(this.out, KeyValueUtil.getSerializedSize(cell, false));
      KeyValueUtil.oswrite(cell, out, false);
    }
  }

  public static class KeyValueDecoder extends BaseDecoder {
    public KeyValueDecoder(final InputStream in) {
      super(in);
    }

    protected Cell parseCell() throws IOException {
      // No tags here
      return KeyValueUtil.iscreate(in, false);
    }
  }

  public static class ByteBuffKeyValueDecoder implements Codec.Decoder {

    protected final ByteBuff buf;
    protected Cell current = null;

    public ByteBuffKeyValueDecoder(ByteBuff buf) {
      this.buf = buf;
    }

    @Override
    public boolean advance() throws IOException {
      if (!this.buf.hasRemaining()) {
        return false;
      }
      int len = buf.getInt();
      ByteBuffer bb = buf.asSubByteBuffer(len);
      if (bb.isDirect()) {
        this.current = createCell(bb, bb.position(), len);
      } else {
        this.current = createCell(bb.array(), bb.arrayOffset() + bb.position(), len);
      }
      buf.skip(len);
      return true;
    }

    @Override
    public Cell current() {
      return this.current;
    }

    protected Cell createCell(byte[] buf, int offset, int len) {
      return new ShareableMemoryNoTagsKeyValue(buf, offset, len);
    }

    protected Cell createCell(ByteBuffer bb, int pos, int len) {
      // We know there is not going to be any tags.
      return new ShareableMemoryOffheapKeyValue(bb, pos, len, false, 0);
    }

    static class ShareableMemoryKeyValue extends KeyValue implements ShareableMemory {
      public ShareableMemoryKeyValue(byte[] bytes, int offset, int length) {
        super(bytes, offset, length);
      }

      @Override
      public Cell cloneToCell() {
        byte[] copy = Bytes.copy(this.bytes, this.offset, this.length);
        KeyValue kv = new KeyValue(copy, 0, copy.length);
        kv.setSequenceId(this.getSequenceId());
        return kv;
      }
    }

    static class ShareableMemoryNoTagsKeyValue extends NoTagsKeyValue implements ShareableMemory {
      public ShareableMemoryNoTagsKeyValue(byte[] bytes, int offset, int length) {
        super(bytes, offset, length);
      }

      @Override
      public Cell cloneToCell() {
        byte[] copy = Bytes.copy(this.bytes, this.offset, this.length);
        KeyValue kv = new NoTagsKeyValue(copy, 0, copy.length);
        kv.setSequenceId(this.getSequenceId());
        return kv;
      }
    }

    static class ShareableMemoryOffheapKeyValue extends OffheapKeyValue implements ShareableMemory {
      public ShareableMemoryOffheapKeyValue(ByteBuffer buf, int offset, int length) {
        super(buf, offset, length);
      }

      public ShareableMemoryOffheapKeyValue(ByteBuffer buf, int offset, int length, boolean hasTags,
          long seqId) {
        super(buf, offset, length, hasTags, seqId);
      }

      @Override
      public Cell cloneToCell() {
        byte[] copy = new byte[this.length];
        ByteBufferUtils.copyFromBufferToArray(copy, this.buf, this.offset, 0, this.length);
        KeyValue kv;
        if (this.hasTags) {
          kv = new KeyValue(copy, 0, copy.length);
        } else {
          kv = new NoTagsKeyValue(copy, 0, copy.length);
        }
        kv.setSequenceId(this.getSequenceId());
        return kv;
      }
    }
  }

  /**
   * Implementation depends on {@link InputStream#available()}
   */
  @Override
  public Decoder getDecoder(final InputStream is) {
    return new KeyValueDecoder(is);
  }

  @Override
  public Decoder getDecoder(ByteBuff buf) {
    return new ByteBuffKeyValueDecoder(buf);
  }

  @Override
  public Encoder getEncoder(OutputStream os) {
    return new KeyValueEncoder(os);
  }
}

File: hbase-native-client/core/keyvalue-codec.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "core/keyvalue-codec.h"

#include <string>

namespace hbase {

KeyValueCodec::KVDecoder::KVDecoder(std::unique_ptr<folly::IOBuf> cell_block, uint32_t offset,
                                    uint32_t length)
    : cell_block_(std::move(cell_block)), offset_(offset), length_(length) {}

KeyValueCodec::KVDecoder::~KVDecoder() {}

std::shared_ptr<Cell> KeyValueCodec::KVDecoder::Decode(folly::io::Cursor &cursor) {
  uint32_t key_length = cursor.readBE<uint32_t>();
  uint32_t value_length = cursor.readBE<uint32_t>();
  uint16_t row_length = cursor.readBE<uint16_t>();
  std::string row = cursor.readFixedString(row_length);
  uint8_t column_family_length = cursor.readBE<uint8_t>();
  std::string column_family = cursor.readFixedString(column_family_length);
  int qualifier_length =
      key_length - (row_length + column_family_length + kHBaseSizeOfKeyInfrastructure_);
  std::string column_qualifier = cursor.readFixedString(qualifier_length);
  uint64_t timestamp = cursor.readBE<uint64_t>();
  uint8_t key_type = cursor.readBE<uint8_t>();
  std::string value = cursor.readFixedString(value_length);

  return std::make_shared<Cell>(row, column_family, column_qualifier, timestamp, value,
                                static_cast<hbase::CellType>(key_type));
}

bool KeyValueCodec::KVDecoder::Advance() {
  if (end_of_cell_block_) {
    return false;
  }

  if (cur_pos_ == length_) {
    end_of_cell_block_ = true;
    return false;
  }

  folly::io::Cursor cursor(cell_block_.get());
  cursor.skip(offset_ + cur_pos_);
  uint32_t current_cell_size = cursor.readBE<uint32_t>();
  current_cell_ = Decode(cursor);
  cur_pos_ += kHBaseSizeOfInt_ + current_cell_size;
  return true;
}

uint32_t KeyValueCodec::KVDecoder::CellBlockLength() const { return length_; }
} /* namespace hbase */
